Replace SplineDefinition with IECore::Ramp, and Store RampData in Shaders

danieldresser-ie opened the pull request:
In combination with the recent Cortex PR: https://github.com/ImageEngine/cortex/pull/1499,
this implements the new way of working with shader ramp parameters that we've been discussing.

There are a lot of changes here, but most are quite trivial - it all follows pretty directly from the Cortex changes. A lot of just renaming things, though there are a few places where I was actually able to cut down on code duplication.

The commit sequence is structured so that after the first 5 commits, this will build against the new Cortex, while having made as few changes as possible ( this was handy when checking behaviour changes ).

The piece I'm most skeptical of is "FIX : ColorRamp and FloatRamp : Fix actual default value", which is why I made it a separate commit ... maybe that's getting too tricksy with the compatibility config? The advantage of this approach is that once all production files have been produced with new Gaffer, we could get rid of the compatibility config, and things would be clean, rather than needing to keep around splineUIMetadata.py for eternity.

johnhaddon commented on python/GafferUI/RampPlugValueWidget.py:
```diff
@@ -34,16 +35,121 @@
 
+	def __init__( self, plug, **kw ) :
+
+		self.__splineWidget = GafferUI.RampWidget()
```
__splineWidget -> __rampWidget?

It would have been nice if this widget refactor was in a separate commit I think - there's more going on than the simple renaming in the rest of the files. The rename of the class will also need calling out in Changes.md.

johnhaddon commented on startup/GafferOSLUI/splineUIMetadata.py:
```diff
@@ -1,40 +0,0 @@
-import Gaffer
-
-Gaffer.Metadata.registerValue( "osl:shader:Pattern/ColorSpline:direction", "userDefault", "v" )
-Gaffer.Metadata.registerValue( "osl:shader:Pattern/FloatSpline:direction", "userDefault", "v" )
```
> The advantage of this approach is that once all production files have been produced with new Gaffer, we could get rid of the compatibility config, and things would be clean, rather than needing to keep around splineUIMetadata.py for eternity.

We can get rid of `splineUIMetadata.py` now, regardless of any other considerations. The `userDefault` registration just affects nodes created by the UI in the future, and we're not letting people create the old shaders in the UI any more.

danieldresser-ie commented on include/Gaffer/RampPlug.h:
```diff
@@ -41,98 +41,22 @@
-/// with one of the interpolations above.
+/// The RampPlug allows the user to manipulate an IECore::Ramp, which is
+/// a simple curve representation with a list of control points and an
+/// interpolations.
```
Addressed as part of: https://github.com/GafferHQ/gaffer/pull/6717/commits/59f2572b279d3ca5cfbf740c3443e2e85f5d801f

danieldresser-ie commented on python/GafferUI/RampWidget.py:
```diff
@@ -63,37 +63,37 @@ def __init__( self, spline=None, drawMode=DrawMode.Splines, **kw ) :
 		self._qtWidget().paintEvent = Gaffer.WeakMethod( self.__paintEvent )
 
-	def setSpline( self, spline ) :
+	def setRamp( self, splineDefinition ) :
```
Addressed in https://github.com/GafferHQ/gaffer/commit/59f2572b279d3ca5cfbf740c3443e2e85f5d801f

danieldresser-ie commented on src/GafferCycles/IECoreCyclesPreview/ShaderNetworkAlgo.cpp:
```diff
@@ -189,8 +189,12 @@ ccl::ShaderNode *convertWalk( const ShaderNetwork::Parameter &outputParameter, c
 		{
+			// \todo : Feels bad that there are no unit tests for this, but currently we don't even bind
+			// convertGraph() to Python, so I don't really want to go to the work of testing it while
+			// in the middle of reworking how we handle ramps.
```
Removed comment in https://github.com/GafferHQ/gaffer/commit/59f2572b279d3ca5cfbf740c3443e2e85f5d801f

danieldresser-ie commented on src/IECoreArnold/CameraAlgo.cpp:
```diff
@@ -88,27 +88,29 @@ AtVector2 curvePoint( const Splineff::Point &point )
 void setShutterCurveParameter( AtNode *camera, const IECore::Data *value, const std::string &messageContext )
 {
-	auto *splineData = runTimeCast<const SplineffData>( value );
+	auto *splineData = runTimeCast<const IECore::RampffData>( value );
```
Addressed in https://github.com/GafferHQ/gaffer/commit/59f2572b279d3ca5cfbf740c3443e2e85f5d801f

danieldresser-ie commented on shaders/Pattern/ColorRamp.osl:
```diff
@@ -0,0 +1,112 @@
+
+#include "GafferOSL/Spline.h"
+
+shader ColorRamp
```
Addressed in 7f6b70f32786642e6f0fbaedea311335b6858454

danieldresser-ie commented:
I think I've now addressed all comments ... hopefully it doesn't bog your browser down too much, since you were saying this PR was already giving you issues for some reason. If it would help, I can try to pull anything that isn't a simple rename out of the big rename commit - though I'd probably do that after squashing 59f2572b279d3ca5cfbf740c3443e2e85f5d801f into the big rename.

danieldresser-ie commented on startup/Gaffer/splinePlugCompatibility.py:
```diff
@@ -1,114 +0,0 @@
-	return getItem
-
-Gaffer.SplineffPlug.__getitem__ = __getitemWrapper( Gaffer.SplineffPlug.__getitem__ )
-Gaffer.SplinefColor3fPlug.__getitem__ = __getitemWrapper( Gaffer.SplinefColor3fPlug.__getitem__ )
```
Put up a first pass PR addressing other early compatibility files: #6735
